### PR TITLE
[AoTI Minifier] UX Improvement

### PR DESCRIPTION
**Pull request by yushangdi**
Summary:
- When a user specify `TORCHINDUCTOR_MAX_AUTOTUNE=1` env variable, we add `config.max_autotune=True` to the generated minifier_launcher
- We should do this to other inductor configs as well in a followup Diff

Currently in dynamo and aoti minifier, if a config is overwritten by an env variable, the config will not show up in the config list in the minifier_launcher.py file. As a result, when running the minifier_launcher, they need to re-apply the same env variable.
 This is:
1) not convenient for the users
2) if they copy-paste the minifier_launcher.py to us without including the env variable, we could be confused and not able to reproduce the error.

Underlying implementation change:

- Add `env_default` parameter to `codegen_config()`. If set, configs overriden by the env are not considered default.

Test Plan:
```
 buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:utils -- -r test_codegen_config
```

Differential Revision: D67299312

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/utils/_config_module.py [hunk 1]
```diff
@@ -432,19 +454,20 @@ def _get_dict(
             skip_default does two things. One if a key has not been modified
                 it skips it. The other is it modified the logging behaviour
                 to match what codegen already did for modified skipped keys
+            env_default: If set, configs overriden by the env are not considered default.
```
Comment (desertfire) [inline torch/utils/_config_module.py:457]: I think this should always be True?

**Comment by desertfire** [comment]
Adding more inductor reviewers

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -293,7 +293,7 @@ def bundle_triton_into_fx_graph_cache_default() -> Optional[bool]:
 inter_node_bw = 25
 
 # enable slow autotuning passes to select algorithms
-max_autotune = os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE") == "1"
+max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")
```
Comment (eellison) [inline torch/_inductor/config.py:296]: Why do we need this ? the existing minifier serializes config settings without this change

**Review by eellison — COMMENTED** [review]
Can you reuse the existing config serialization from minifier? i would not expect this many changes

File: torch/_inductor/config.py (hunk 2 above)
Comment (yushangdi) [reply torch/_inductor/config.py:296]: @eellison 

The problem is that the minifier cannot produce the same config as the run time config.

For example, if you run the script below like

If you just run the code below directly, like `TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py`, the configs that the minifier dumps doesn't have `torch._inductor.config.max_autotune = True`, even though you have env variable `TORCHINDUCTOR_MAX_AUTOTUNE=1`.


This is because the config module takes the ` os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE") == "1"` as default and skipped this config, which is wrong. The true default value is False, and the user actually changed the value of this config with an env variable. 

```
import torch
from torch._inductor import config as inductor_config
class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = torch.nn.Linear(10, 16)
        self.relu = torch.nn.ReLU()
        self.sigmoid = torch.nn.Sigmoid()
    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.sigmoid(x)
        return x
inductor_config.aot_inductor.dump_aoti_minifier = True
torch._inductor.config.triton.inject_relu_bug_TESTING_ONLY = "compile_error"
with torch.no_grad():
    model = Model().to("cuda")
    example_inputs = (torch.randn(8, 10).to("cuda"),)
```

Comment (yushangdi) [reply torch/_inductor/config.py:296]: Most of the changes here is to add a flag when dumping the configs, so we don't skip the configs that are changed by the setting the env variables.

File: torch/_dynamo/config.py [hunk 3]
```diff
@@ -421,6 +421,14 @@ def default_debug_dir_root():
     if DEBUG_DIR_VAR_NAME in os.environ:
         return os.path.join(os.environ[DEBUG_DIR_VAR_NAME], "torch_compile_debug")
     elif is_fbcode():
+        fbcode_dir = os.path.join(os.path.expanduser("~"), "fbsource", "fbcode")
```
Comment (jansel) [inline torch/_dynamo/config.py:424]: Is this path general?  Should this be cwd or home?  Putting it in the fbsource tree seems a bit weird.

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -1380,3 +1380,15 @@ class test_configs:
...
+#     """
+#     Set config to True if env is set to 1.
+#     We use the if statement to avoid calling setattr if env is not set.
+#     This helps config.codegen_config() to only generate configs that are not
+#     the default value.
+#     """
+#     if os.environ.get(env) == "1":
+#         setattr(sys.modules[__name__], config, True)
+
+# _set_config_from_env("max_autotune", "TORCHINDUCTOR_MAX_AUTOTUNE")
```
Comment (jansel) [inline torch/_inductor/config.py:1394]: Dead code?

File: torch/utils/_config_module.py [hunk 5]
```diff
@@ -407,14 +407,36 @@ def _set_alias_val(self, entry: _ConfigEntry, val: Any) -> None:
...
+                and not_set_env_force
+            )
         return self._config[name].user_override is _UNSET_SENTINEL
 
     def _get_dict(
         self,
         ignored_keys: Optional[List[str]] = None,
         ignored_prefixes: Optional[List[str]] = None,
         skip_default: bool = False,
+        env_default: bool = False,
```
Comment (jansel) [inline torch/utils/_config_module.py:439]: Does anything break if we just turn this on always?  If so let's just remove the option and make it default.

File: torch/_dynamo/repro/aoti.py [hunk 6]
```diff
@@ -455,7 +455,7 @@ def common_flags(parser):
         )
 
     subparsers = parser.add_subparsers(
-        dest="command", metavar="{run,minify,analyze}", required=True
+        dest="command", metavar="{run,minify}", required=True
```
Comment (jansel) [inline torch/_dynamo/repro/aoti.py:458]: Was this unused?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/config.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/config.py:296]: I think we should be able to do the above without the `max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")` changes

Comment (yushangdi) [reply torch/_inductor/config.py:296]: > I think we should be able to do the above without the `max_autotune = Config(False, env_name_default="TORCHINDUCTOR_MAX_AUTOTUNE")` changes.. But I guess this is the standard way, so fine with me..

@eellison yeah I agree, we could do it another way, but since there is the `Config` class already, I thought we might as well just use it.

File: torch/_dynamo/repro/aoti.py (hunk 6 above)
Comment (yushangdi) [reply torch/_dynamo/repro/aoti.py:458]: yeah. the AoTI minifier doesn't have an analyze mode yet.

File: torch/utils/_config_module.py (hunk 5 above)
Comment (yushangdi) [reply torch/utils/_config_module.py:439]: @c00w Do you know if we can make this `env_default` default to True? (If a config's env value is different from default value, it cannot be skipped by `skip_default`).

Currently, the test is written such that these configs are skipped by `skip_default`.

File: torch/utils/_config_module.py (hunk 1 above)
Comment (yushangdi) [reply torch/utils/_config_module.py:457]: Currently, the test is written such that these configs are skipped by skip_default. If we change this to always be True, we need to change the existing tests. I'm checking with test writer (Colin) if we can change it.

File: torch/_dynamo/config.py (hunk 3 above)
Comment (yushangdi) [reply torch/_dynamo/config.py:424]: It's intended to be in the fbcode. The generated TARGETS file and python files need to be in fbsource for them to be compiled by buck.

Comment (yushangdi) [reply torch/_dynamo/config.py:424]: I guess I could also make it cwd.

File: torch/utils/_config_module.py (hunk 5 above)
Comment (yushangdi) [reply torch/utils/_config_module.py:439]: I think it's probably ok. I'll make it default then.

File: torch/utils/_config_module.py (hunk 1 above)
Comment (yushangdi) [reply torch/utils/_config_module.py:457]: I'll change it to always be True

File: torch/_dynamo/repro/aoti.py (hunk 6 above)
Comment (yushangdi) [reply torch/_dynamo/repro/aoti.py:458]: yeah, AoTI minifier doesn't have an analyze mode yet

**Comment by yushangdi** [comment]
@jansel @desertfire Pin for review on the updated version of the PR. Thank you!

File: torch/_inductor/config.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/config.py:299]: It does work, but I would still prefer we do the general thing that covers all configs instead of just this specific one.

**Review by jansel — APPROVED** [review]

**Review by eellison — APPROVED** [review]
🚀 🚀 🚀

**Comment by yushangdi** [comment]
> This pull request has been **reverted** by [66ce13b](https://github.com/pytorch/pytorch/commit/66ce13b4974ad7b470f60bf871604bff1c87f4ec). To re-land this change, please open another pull request, assignthe same reviewers, fix the CI failures that caused the revert and make sure that the failing CI runs on the PR by applying the proper ciflow label (e.g., ciflow/trunk).

Note: this Diff is only partially reverted in [66ce13b]. The majority of the Diff is not reverted.
